server/routes/events.js

```diff
@@ -12,120 +13,447 @@ router.get('/ping', function(req, res, next) {
+				await host_query.transacting(trx);
+				return trx.commit;
+			})
+	})
+	.then(function() {
+		res.send(util.message('Successfully inserted new event: ' + req.body.name));
+	})
+	.catch(err => { return next(err) });
+});
+
+// Updated a single event by event_id
+router.put('/:event_id', function(req, res, next) {
+	const event_id = req.params.event_id;
+	let validReqForEvent = false;
+
+	let values = {
+  	name: req.body.name, 
+  	starts_at: req.body.starts_at, 
+  	ends_at: req.body.ends_at,
+  	quarter: req.body.quarter,
+  	location_id: req.body.location_id,
+  	description: req.body.description,
+  	fb_event: req.body.fb_event,
+  	picture: req.body.picture,
+  	is_featured: req.body.is_featured
+  };
+
+  let category_ids = req.body.categories;
+	let category_values = [];
+  let company_ids = req.body.companies;
+	let company_values = [];
+  let host_ids = req.body.hosts;
+  let host_values = [];
+
+  if (category_ids){
+	  category_ids.forEach(function(entry) {
+	  	category_values.push({ event_id: event_id, category_id: entry });
+	  });
+	}
+
+  if (company_ids){
+	  company_ids.forEach(function(entry) {
+	  	company_values.push({ event_id: event_id, company_id: entry });
+	  });
+	}
+
+  if (host_ids){
+	  host_ids.forEach(function(entry) {
+	  	host_values.push({ event_id: event_id, host_id: entry });
+	  });
+	}
+
+	let remove_category_ids = req.body.remove_categories;
+	let remove_company_ids = req.body.remove_companies;
+  let remove_host_ids = req.body.remove_hosts;
+
+  Object.keys(values).forEach(function(key) {
+  	if (!validReqForEvent && values[key]) {
+  		validReqForEvent = true;
+  	}
+	});
+
+	var event_query = knex('event').modify(function(queryBuilder) {
+		if (validReqForEvent){
+			queryBuilder.where({ fb_id: event_id }).update(values);
+		}
+	});
+
+	var remove_category_query = knex('event_category').modify(function(queryBuilder) {
+		if (remove_category_ids){
+			queryBuilder.where({ event_id: event_id })
+				.andWhere((builder) =>
+					builder.whereIn('category_id', remove_category_ids)
+				)
+				.del();
+		}
+	});
+
+	var remove_host_query = knex('event_host').modify(function(queryBuilder) {
+		if (remove_host_ids){
+			queryBuilder.where({ event_id: event_id })
+				.andWhere((builder) =>
+					builder.whereIn('host_id', remove_host_ids)
+				)
+				.del();
+		}
+	});
+
+	var remove_company_query = knex('event_company').modify(function(queryBuilder){
+		if (remove_company_ids){
+			queryBuilder.where({ event_id: event_id })
+				.andWhere((builder) =>
+					builder.whereIn('company_id', remove_company_ids)
+				)
+				.del();
+		}
+	});
+
+	var category_query = knex('event_category').insert(category_values);
+	var host_query = knex('event_host').insert(host_values);
+	var company_query = knex('event_company').insert(company_values);
+
+	knex.transaction(function(trx) {
+		return event_query.transacting(trx)
+			.then(async function() {
+				await category_query.transacting(trx);
```

Comment:
We don't want to do all these queries every time, do we? They will throw an error if no values are present to insert/remove whatever. Wrap these in if statements or some sort of conditional check

Reply:
The .modify thing from ^ is for the if statements!

Reply:
So even if there are no hosts passed as parameters, the line "await host_query.transacting(trx);" won't throw an error?

Reply:
Yeah, it's all good!